### DISCUSSION TITLE
Question about handling of memory layout

### DISCUSSION THREAD

**Discussion by SeleDreams** [rust3ds/ctru-rs#185] [Q&A]
Hi,
This issue is more about questions on how ctru-rs works.
I am working on a wrapper of libnds for the DS and I would like to be able to do something like ctru-rs that can simply be included as a dependency to build for ds with just the .json of the target.

The DS and 3DS both work with specs files that define .ld, .mem files that are included during the build process.
I tried to link them with `println!("cargo:rustc-link-arg=-specs={}/sys/crts/ds_arm9.specs",&blocksds_path);` for the DS but i found out that build.rs link args don't extend to parent crates, as such this wouldn't work when used as a dependency.

How does ctru-rs handle this type of things ? as from what I know it depends on devkitpro which does rely on specs files as well

**Comment by Meziu** [comment]
These are things that cannot be handled by a simple build script, since they override compilation on a compiler/linker level.

You can specify them as additional linker arguments when creating a target triple `.json` file or `rustc-target` configuration.

[Here](https://github.com/rust-lang/rust/blob/master/compiler/rustc_target/src/spec/targets/armv6k_nintendo_3ds.rs#L10)'s where we specify these requirements.

Admittedly, this is the only direct reference of `devkitPro` specific configurations we make in that target triple, but we have also put the toolchain as a requirement in the target docs, so it's fine.

**Reply by SeleDreams** [reply]
I originally tried adding them to the json, the issue is that the json does not support environment variables.
I guess the fact you added them directly to rust's source code helps

what do you mean by rustc-target configuration ?

**Reply by Meziu** [reply]
> I originally tried adding them to the json, the issue is that the json does not support environment variables.

Which is why we actually rely on the global path for these tools (still in a rough spot for that).

> I guess the fact you added them directly to rust's source code helps

From the Rust code we could avoid issues by using `env`, but I always decided against that. (I am very much changing my mind)

> what do you mean by rustc-target configuration ?

Basically the target-triple module we pushed to `rustc_target` (which is the compiler's crate that holds information for all of its targets).

It is functionally equal to the target triple `json` file.

**Reply by SeleDreams** [reply]
I see
global path wise, is there a way to set it somewhere ? because if I just do -specs=ds_arm9.specs it won't find it unless I like copy the specs file in my project directory

**Reply by Meziu** [reply]
> I see
> 
> global path wise, is there a way to set it somewhere ? because if I just do -specs=ds_arm9.specs it won't find it unless I like copy the specs file in my project directory

I find weird that your linker can't find the file, since it should be placed in a relative file directory to which it has direct access. Are you sure the linker running is the one by devkitPro?

Usually we had problems of not finding the linker (on the global path), but once found it was always able to find its files.

**Reply by SeleDreams** [reply]
I am using BlocksDS instead of devkitpro (a newer toolchain for the DS with improvements)
https://blocksds.github.io/docs/introduction/introduction/

**Reply by Meziu** [reply]
> I am using BlocksDS instead of devkitpro (a newer toolchain for the DS with improvements)
> 
> https://blocksds.github.io/docs/introduction/introduction/

Ouch, this makes the situation a bit harder. Are you sure the file is properly included? I have no idea on how their compiler toolchain works.

**Reply by SeleDreams** [reply]
I was able to make it work by forking cargo-3ds, modifying it for ds specific stuff, and inserting the specs file with the correct path based on the BLOCKSDS environment variable.
So I now have a working cargo-nds to build ds homebrews. though at the moment, only arm9 homebrews are supported (which should be the majority of cases, but I still plan to support arm7 long term)

https://github.com/SeleDreams/cargo-nds

**Comment by SeleDreams** [comment]
Since I am in the same discord as the main BlocksDS developer, I'll talk to
them about it. They might find a way to make things easier

**Comment by SeleDreams** [comment]
I figured out the best way to handle it with the limitations.
I'm going to try, in the new cargo-nds I'll make to make it create a
symlink to the specs file's directory before the build so that the linker
can find it from the project's relative path. And then it would delete the
symlink after the build.
That seems like the best way to handle it
